### ISSUE TITLE
[bug] Warmup LR is one step behind: OptimizerParamScheduler not advanced before the first optimizer step

### ISSUE THREAD

**Issue by anveshjayakrishna** [NVIDIA-NeMo/Megatron-Bridge#5399]
### Summary
  
  Megatron-Bridge steps the optimizer before advancing the LR scheduler, and the
  scheduler is not advanced once at construction. As a result iteration 0 runs the
  optimizer at the warmup **init** LR (`num_steps = 0`), and warmup stays one
  increment behind for the rest of the warmup phase.
  
  ### Details
  
  The training loop steps the optimizer first, then the scheduler:
  
  - [`train.py#L926`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/training/train.py#L926) — `optimizer.step()`
  - [`train.py#L954`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/training/train.py#L954) — `scheduler.step(increment=increment)`
  
  And the scheduler is constructed and returned without an initial step:
  
  - [`optim.py#L339`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/training/optim.py#L339) — `OptimizerParamScheduler(...)` is built and returned as-is; `_get_scheduler` never calls `.step()`.
  
  So at iteration 0 the scheduler is still at `num_steps = 0` (LR = `init_lr`). The
  first `optimizer.step()` runs at that init LR, and because the first
  `scheduler.step()` happens only afterward, warmup is offset by one increment
  throughout.
  
  ### Why NeMo (Lightning) is not affected
  
  NeMo's Lightning schedulers extend PyTorch's `_LRScheduler`: `WarmupPolicy.__init__`
  calls `super().__init__(optimizer, last_epoch)`, and `_LRScheduler.__init__` runs
  `_initial_step()` → `self.step()`
  ([torch/optim/lr_scheduler.py#L149-L154](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/optim/lr_scheduler.py#L149-L154)).
  That construction-time step applies the step-0 warmup LR to the optimizer *before*
  the first `optimizer.step()`, so NeMo's warmup starts one increment ahead — and
  its loop then just keeps advancing it each step.
  
  Megatron-Bridge's `OptimizerParamScheduler` does **not** extend `_LRScheduler`, so
  it never gets that construction-time step. Building it directly and stepping it 
  only inside the loop is what leaves warmup one increment behind an
  `_LRScheduler`-based setup like NeMo's.
  
  ### Question for maintainers
  
  Is the one-step warmup offset intentional (treating iteration 0 as a zero-LR
  warmup step by design), or should the scheduler be advanced once before the first
  `optimizer.step()` so warmup starts at the first nonzero increment? Happy to send
  a PR for whichever behavior is preferred.

### Minimal repro

```shell
Inside train.py's per-iteration train_step, the two relevant calls run in this order:

  # train.py#L926
  update_successful, grad_norm, num_zeros_in_grad = optimizer.step()
  ...
  # train.py#L954
  scheduler.step(increment=increment)

  At iteration 0, optimizer.step() (L926) runs before the very first scheduler.step() (L954). Since setup_optimizer returns the scheduler with no initial step (optim.py#L339, num_steps = 0), the optimizer's param-group LR is still init_lr (0.0) at that first L926 call. The first
  nonzero warmup LR is only applied at L954 — one step too late — so every warmup iteration thereafter is offset by one increment.

  Minimal check without a full run — reproduce the exact L926/L954 ordering against Bridge's scheduler:

  from megatron.bridge.training.optim import setup_optimizer
  optimizer, scheduler = setup_optimizer(optimizer_config, scheduler_config, model)
  # ... with SchedulerConfig(lr_warmup_init=0.0, lr_warmup_steps=10, ...)

  # What L926 (optimizer.step) sees on iteration 0, before L954 runs:
  print(optimizer.param_groups[0]["lr"])   # -> 0.0

  scheduler.step(increment=1)               # L954 for iteration 0
  print(optimizer.param_groups[0]["lr"])    # -> first warmup increment (e.g. 1e-4)
```

### Expected behavior

 When optimizer.step() at L926 runs on iteration 0, the param-group LR should already be the step-0 warmup increment, not init_lr — i.e. setup_optimizer should hand back a scheduler that has taken its initial step (or the loop should advance the scheduler before the first L926).
  Then the LR applied at L926 matches what an _LRScheduler-based scheduler applies at construction, and warmup is no longer offset by one.

### Affected area

area:training

### Regression?

Not sure

### Logs

```shell

```

**Comment by adityavavreNVDA** [comment]
Thanks for the detailed writeup. The observation is correct, but this looks intended rather than a missing call.

The scheduler does take an initial step inside MCore (`megatron/core/optimizer_param_scheduler.py`) rather than in Bridge's `_get_scheduler`:

```python
# Set the learning rate
self.step(0)
```

That sets param_group['lr'] to get_lr() at num_steps = 0, which the warmup branch evaluates to init_lr. Bridge passes `init_lr=scheduler_config.lr_warmup_init` (optim.py).

So iteration 0 runs at `lr_warmup_init` which is the configured value for it, not a stale or unset LR. If you don't want a zero-LR first step, that's the knob to change. If there are no other concerns I will close this issue.

**Comment by huvunvidia** [comment]
Hi @anveshjayakrishna ,
Just check-in to see if [adityavavreNVDA](https://github.com/adityavavreNVDA) 's answer resolves your issue.
If so (or in case you want to leave this issue for later), we can close it.
